Source/WebKit/UIProcess/Inspector/WebFrameInspectorTargetProxy.cpp

```diff
@@ -0,0 +1,111 @@
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WebFrameInspectorTargetProxy.h"
+
+#include "InspectorTargetProxy.h"
+#include "MessageSenderInlines.h"
+#include "ProvisionalFrameProxy.h"
+#include "WebFrameMessages.h"
+#include "WebFrameProxy.h"
+#include <JavaScriptCore/InspectorTarget.h>
+#include <memory>
+#include <wtf/TZoneMallocInlines.h>
+
+namespace WebKit {
+
+using namespace Inspector;
+
+WTF_MAKE_TZONE_ALLOCATED_IMPL(WebFrameInspectorTargetProxy);
+
+std::unique_ptr<WebFrameInspectorTargetProxy> WebFrameInspectorTargetProxy::create(WebFrameProxy& frame, const String& targetId)
+{
+    return makeUnique<WebFrameInspectorTargetProxy>(frame, targetId);
+}
+
+std::unique_ptr<WebFrameInspectorTargetProxy> WebFrameInspectorTargetProxy::create(ProvisionalFrameProxy& provisionalFrame, const String& targetId)
+{
+    Ref frame { provisionalFrame.frame() };
+    std::unique_ptr target = WebFrameInspectorTargetProxy::create(frame.get(), targetId);
+    target->m_provisionalFrame = provisionalFrame;
+    return target;
+}
+
+WebFrameInspectorTargetProxy::WebFrameInspectorTargetProxy(WebFrameProxy& frame, const String& targetId)
+    : InspectorTargetProxy(targetId, Inspector::InspectorTargetType::Frame)
+    , m_frame(frame)
+{
+}
+
+WebFrameInspectorTargetProxy::~WebFrameInspectorTargetProxy() = default;
+
+void WebFrameInspectorTargetProxy::connect(Inspector::FrontendChannel::ConnectionType connectionType)
+{
+    if (RefPtr provisionalFrame = m_provisionalFrame.get()) {
+        provisionalFrame->protectedProcess()->send(Messages::WebFrame::ConnectInspector(connectionType), provisionalFrame->protectedFrame()->frameID());
+        return;
+    }
+
+    Ref frame = m_frame.get();
+    frame->protectedProcess()->send(Messages::WebFrame::ConnectInspector(connectionType), frame->frameID());
+}
+
+void WebFrameInspectorTargetProxy::disconnect()
+{
+    if (isPaused())
+        resume();
+
+    if (RefPtr provisionalFrame = m_provisionalFrame.get()) {
+        provisionalFrame->protectedProcess()->send(Messages::WebFrame::DisconnectInspector(), provisionalFrame->protectedFrame()->frameID());
+        return;
+    }
+
+    Ref frame = m_frame.get();
+    frame->protectedProcess()->send(Messages::WebFrame::DisconnectInspector(), frame->frameID());
+}
+
+void WebFrameInspectorTargetProxy::sendMessageToTargetBackend(const String& message)
+{
+    if (RefPtr provisionalFrame = m_provisionalFrame.get()) {
```

Comment:
Okay, I have a design question. is there any relationship between m_frame and m_provisionalFrame, or are they being used mutually exclusively? I know that this mirrors the PageTarget code that already exists, I'm just not sure if we are calling disconnect() on the provisional frame target before the m_provisionalFrame gets nulled out.

Alternatively, can you add a bugzilla link to track adding more tests for provisional frame targets? I think my questions would be answered by the test case.